bdb/file.c

```diff
@@ -7662,159 +7664,156 @@ static int bdb_process_unused_files(bdb_state_type *bdb_state, tran_type *tran,
+        file_name_post_prefix = ent->d_name + tp_len;
 
-            /* brute force scan to find any files on disk that we aren't
-             * actually using */
-            if (file_version) {
-                int found_in_llmeta = 0;
-                int i;
+        /* try to parse a file version */
+        invers = strtoull(file_name_post_prefix, &endp, 16 /*base*/);
 
-                /* We have to check new. prefix for schemachange first.
-                 * See NOTE in bdb_is_new_sc_file()
-                 */
-                rc = bdb_is_new_sc_file(bdb_state, tran, bdb_state->name,
-                                        file_version, bdberr);
-                if (rc == 1) {
+        /* if no file_version was found after the prefix or the next thing
+         * after the file version isn't .blob or .data or .index */
+        if (endp == file_name_post_prefix ||
+            (strncmp(endp, blob_ext, sizeof(blob_ext) - 1) != 0 &&
+             strncmp(endp, data_ext, sizeof(data_ext) - 1) != 0 &&
+             strncmp(endp, qdb_ext, sizeof(qdb_ext) - 1) != 0 &&
+             strncmp(endp, index_ext, sizeof(index_ext) - 1) != 0)) {
+            file_version = 0;
+        } else {
+            uint8_t *p_buf = (uint8_t *)&invers,
+                    *p_buf_end = p_buf + sizeof(invers);
+            struct bdb_file_version_num_type p_file_version_num_type;
+
+            bdb_file_version_num_get(&p_file_version_num_type, p_buf,
+                                     p_buf_end);
+
+            file_version = p_file_version_num_type.version_num;
+        }
+
+        if (file_version)
+            continue;
+
+        /* brute force scan to find any files on disk that we aren't
+         * actually using */
+        int found_in_llmeta = 0;
+        int i;
+
+        /* We have to check new. prefix for schemachange first.
+         * See NOTE in bdb_is_new_sc_file()
+         */
+        rc = bdb_is_new_sc_file(bdb_state, tran, bdb_state->name, file_version,
+                                bdberr);
+        if (rc == 1) {
+            found_in_llmeta = 1;
+            rc = 0;
+        } else if (rc) {
+            logmsg(LOGMSG_ERROR,
+                   "%s:%d failed to check llmeta for %s, rc %d, bdberr "
+                   "%d\n",
+                   __func__, __LINE__, ent->d_name, rc, *bdberr);
+            continue;
+        }
+
+        /* try to find the file version amongst the active data files */
+        for (i = 0; i < bdb_state->numdtafiles; ++i) {
+            if (bdb_state->bdbtype == BDBTYPE_QUEUEDB) {
+                rc = bdb_get_file_version_qdb(bdb_state, tran, &version_num,
+                                              bdberr);
+            } else {
+                rc = bdb_get_file_version_data(bdb_state, tran, i, &version_num,
+                                               bdberr);
+            }
+            if (rc == 0 && version_num == file_version) {
```

Comment:
I just figured out that this is also wrong.. We dont want to break if rc == 0..

Reply:
That would be behavior different from original. It seems that we declare found_in_llmeta = 1 if rc==0 and if version retrieved matched.  I did not write the original but it looks correct.

Reply:
I am saying the case where if rc == 0 but version_num != file_version, your patch will break out of the for loop but we should continue instead.

Reply:
try create a table with blob, new code will mistakenly delete all blob files.

Reply:
scindex catches it occasionally because of missing files. But I think we should really have a dedicated test for file deletion.

Reply:
You are a genius for catching this!